Framework/src/RRNG.cc

```diff
@@ -0,0 +1,131 @@
+#include "PandaTree/Framework/interface/RRNG.h"
+#include "Math/QuantFuncMathCore.h"
+#include "TString.h"
+#include <iostream>
+#include <stdexcept>
+#include <cassert>
+
+using namespace std;
+using namespace panda;
+
+int RRNG::_default_idx;
+
+RRNG::RRNG(int seed, int maxEntropy): 
+  _rng(seed),
+  _maxEntropy(maxEntropy)
+{
+  assert(_maxEntropy > 0);
+  _default_idx = -1;
+  _x = new double[_maxEntropy];
+  _available = new bool[_maxEntropy];
+  generate();
+}
+
+RRNG::RRNG(RRNG const& other): 
+  _rng(other._rng),
+  _maxEntropy(other._maxEntropy)
+{
+  _x = new double[_maxEntropy];
+  _available = new bool[_maxEntropy];
```

Comment:
We can also remake `_x, _available` into a vector of `<double, bool>` pairs, so you can iterate over them together with a fancy `iterator` for both initialization and fetching.

Reply:
Wait, I definitely don't think vector resizing is the right thing to do. You can easily use an arbitrary amount of memory by doing big-N calls to `RRNG::poisson`

Reply:
And I guess that's something we ought to discuss... do we want the copy and move constructors (forgot to implement move) to create a new copy with:
- exactly the same random numbers currently as the original
- the same seed/rng/settings, but generate a fresh set of random numbers

I picked the latter option for copy. I suppose the implicitly-defined move constructor is actually going to have the former behavior. And I implemented a copy constructor so that it could be used in `EventBase`'s copy constructor, which I believe is a pretty common use case.

Reply:
Ah true. I forgot about the `EventBase`. Resizing may not be necessary either. But if the settings are the same, why make the copy generate the same numbers again?

Reply:
I think resizing is dangerous. 

Copy is not generating the same numbers again - it's generating `_maxEntropy` numbers from a copy of the `TRandom3` instance, which copies the state of the original rng:

```root [0] TRandom3 a(123);
root [1] a.Rndm()
(Double_t) 6.964692e-01
root [2] TRandom3 b(a);
root [3] b.Rndm()
(Double_t) 7.129553e-01
root [4] a.Rndm()
(Double_t) 7.129553e-01
```

So the behavior would be something like:
```C++
RRNG a(2);     // say it generates {0.4, 0.5}
RRNG b(a);     // say b generates {0.6, 0.7}
a.generate();   // now a contains {0.6, 0.7} too
```

Now that I think about it, maybe we do want `b` to have the same state as `a` on copy... In that case, I think I'll switch to vectors just to take advantage of the of easy copies. I still don't think dynamic resizing is a good idea though. Safer to just re-use the memory.

Reply:
Yeah, I think I agree now on the resizing bit.